### ISSUE TITLE
Quote posts

### ISSUE THREAD

**Issue by jamietanna** [snarfed/bridgy-fed#461]
Quoting a post on Mastodon ends up looking a bit odd (https://indieweb.social/@www.jvt.me@www.jvt.me/110084066464382647) as the original content doesn't get presented - is there something we can do to ie add the link to the post being quoted, rather than trying to embed it?

**Comment by snarfed** [comment]
Hmm, interesting question!

Afaik Mastodon deliberately has no native quote feature, by design. Background in eg https://github.com/mastodon/mastodon/issues/20673, among others. We could try to construct it ourselves, manually, with text and minimal HTML inside content, but I'm always reluctant to do that kind of thing.

I wonder how other fediverse software supports quoting, if at all, that might be a good place to start.

BF currently converts mf2 `quotation-of` to AS2 `attachment`. [Here's the log for this post.](https://fed.brid.gy/log?start_time=1679749575&key=https://www.jvt.me/mf2/2023/03/bepvv/&module=) [Mastodon's AP docs say `attachment` is for media](https://docs.joinmastodon.org/spec/activitypub/#properties-used), and don't say anything about quotes, nor do the [AS2 core](https://www.w3.org/TR/activitystreams-core/) or [vocab](http://www.w3.org/TR/activitystreams-vocabulary/) specs.

**Comment by snarfed** [comment]
Sounds like the semi-standard way to do AP quotes is [FEP-e232](https://socialhub.activitypub.rocks/t/fep-e232-object-links/2722) (https://codeberg.org/fediverse/fep/issues/14, https://socialhub.activitypub.rocks/t/announce-activity/3129/4 ). Looks like it's a `Note` with a `Link` `tag` and content in `content`, eg:

```json


{
    "type": "Note",
    "content": "This is a quote:<br>RE: https://example.com/objects/123",
    "tag": [
        {
            "type": "Link",
            "mediaType": "application/ld+json; profile=\"https://www.w3.org/ns/activitystreams\"",
            "href": "https://example.com/objects/123",
            "name": "RE: https://example.com/objects/123"
        }
    ]
}
```

I assume the `RE: https://example.com/objects/123` part of the content is optional.

**Comment by trwnh** [comment]
the `RE: <url>` is the part that gets replaced by `tag[0]`. basically `name` is the microsyntax. if you wanted to have explicit affordance for "quote tweet" style semantics, you can use `rel`.

for example, foundkey does this: https://akkoma.dev/FoundKeyGang/FoundKey/pulls/318 and expects at least one of the following rels:

- `https://misskey-hub.net/ns#_misskey_quote`
- `http://fedibird.com/ns#quoteUri`
- `https://www.w3.org/ns/activitystreams#quoteUrl`

basically, at some point these three uris were used as JSON-LD properties for the same thing.

**Comment by snarfed** [comment]
Oh wow interesting, microsyntax. Whee, and ugh. Thank you!

**Comment by snarfed** [comment]
This is officially on Mastodon's roadmap: https://github.com/mastodon/mastodon/issues/309, https://joinmastodon.org/roadmap

**Comment by jamietanna** [comment]
Quotes are coming in Mastodon 4.4 - https://oisaur.com/@renchap/112299860209222424

**Comment by temtemy** [comment]
The URL to the quoted post does get included when it comes to Misskey quote posts, though for some reason the URL is not a link. And they should be shown as quote posts in Bluesky anyway because it natively supports it.

Example: https://bsky.app/profile/mima.makai.chaotic.ninja.ap.brid.gy/post/3krmzapfowqg2
Original Misskey post: https://makai.chaotic.ninja/notes/9sv2d3tpar

**Comment by qazmlp** [comment]
This seems to be because Misskey doesn't attach a `"Link"` `"tag"` for the quote URL.
Which also explains why some clients fail to open these quoted posts natively, in my experience.

(Bridgy probably shouldn't require a `"tag"` for turning an `<a>` into a link, just for rewriting to a bridged object where available.)

Misskey instead attaches the quote URL under the canoncial IRI `"https://misskey-hub.net/ns#_misskey_quote"` directly to the post to enable quote presentation, which is documented here: https://misskey-hub.net/ns/

(I wouldn't use `"https://www.w3.org/ns/activitystreams#quoteUrl"`.
That appears to be undocumented and is a potentially collision-prone way to extend the protocol.)

**Comment by temtemy** [comment]
Btw quote posts from Bluesky to AP also seem to fail, I don't see any indication of a quote whether it's Mastodon or Misskey.

Mastodon example: https://yamabi.co/@makai.chaotic.ninja@bsky.brid.gy/112278586683792932
Misskey example: https://makai.chaotic.ninja/notes/9s5ccwqg2f
Original Bluesky quote post: https://bsky.app/profile/did:plc:cu33jkvtkoxyfd45wg7hrhra/post/3kq7qpwkces2q

**Comment by AtiusAmy** [comment]
I feel like Misskey quotes support would be nice as Threads added that as well!

**Comment by snarfed** [comment]
Note to myself: Bluesky quote post `app.bsky.embed.record`s aren't yet implemented in [granary's](https://github.com/snarfed/granary) `bluesky.py`. Need to do that first here. Would be a good starter project for anyone else interested in contributing!

**Comment by wakest** [comment]
This seems like it should get some priority cause how you have it now the feeds with people who are Quote posting a lot totally don't make any sense cause you don't have ANY indication of that they are talking about something else. No need to wait for mastodon to do anything, just include the url of the post they QP about

**Comment by npdoty** [comment]
This would be a huge help for those of us following bridged Bluesky accounts from the fediverse. As it is, many posts simply don't make any sense.

Some Mastodon clients are already handling some version of quoting, even if it's not a part of the core Mastodon software, just by having a URL, and then doing link preview rendering. It's definitely effective, and would make the bridged messages a lot less confusing.

**Comment by AtiusAmy** [comment]
We could also do something like Threads. Threads will have a link to the quote post, but if you're on a software with quote post (for ex: Misskey) it will show as quotes. On software that dont like Mastodon will show as link and if someone quote post a post from someone who is not bridged it will also be a link no matter what software you're on.

English is not my first language, so i must apologize if my point seem not concise

(PS: I also kinda want quote post bridging from AP -> Bsky, I'm from Misskey (one of the many software that supports it) and it looks weird on bsky)

![Screenshot_20240615-044947_Firefox.png](https://github.com/snarfed/bridgy-fed/assets/136548767/2e82fd2c-9b86-4a8c-a699-e8732252056b)

![Screenshot_20240615-045049_Aria.png](https://github.com/snarfed/bridgy-fed/assets/136548767/e672b81c-a280-498f-9f87-ac0644fad6f0)

**Comment by snarfed** [comment]
Progress. Here's how Bluesky quote posts now look in Mastodon (https://bsky.app/profile/snarfed.bsky.social/post/3kv6hsw44wk2q => https://indieweb.social/@snarfed.bsky.social@bsky.brid.gy/112636479451138698 ):

<img width="611" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/7e087956-1102-4a82-9864-b0346aa9018b">

**Comment by snarfed** [comment]
I'm still looking for a Misskey instance I can create an account on to test this. I've looked at https://misskey.io/ and https://quietplace.xyz/ , and neither of them can see @bsky.brid.gy@bsky.brid.gy, so I can't enable the bridge on them. makai.chaotic.ninja, which @temtemy mentioned above, doesn't have open signups. 

@AtiusAmy, any thoughts on where I can test this?

**Comment by snarfed** [comment]
Ah, thank you! And on quietplace.xyz : https://quietplace.xyz/@bsky.brid.gy@bsky.brid.gy

Any idea why it doesn't show up in user search on those instances? Or on eg the users tab of https://quietplace.xyz/instance-info/bsky.brid.gy ?

**Comment by snarfed** [comment]
Useful thoughts from @qazmlp in https://github.com/snarfed/granary/commit/3fc6eb98f28b42f362e0a671d5a4a91d1504676d#commitcomment-143259713 :

> Since posts can be edited, the pointed to post may change from the state it was quoted in. Bluesky appears to have laid some groundwork to detect this (I assume to display a notice in such cases, once editing lands and content can diverge).
> 
> Fediverse instances (unfortunately) don't do this, and it was only mentioned once in FED-e232: Object Links. The proposal linked there was also closed, so I'm just going to assume it's both not planned and, if it was, would look a bit more complex than what Bluesky is doing 🤷
> 
> The alternative is to detect and notify users when a post they've quoted changes, but that's neither reliable in a federated system nor would it work for quoters who don't use social media enough to react. Oh well. Someone might eventually go ahead and do it if it becomes a serious-enough problem.

It seems like current inbox delivery audience targeting could easily be extended to send post `Update`s to all known quote post authors, which could help with this. I don't know if any fediverse projects currently do that, though. I'm guessing now.

**Comment by hybridhavoc** [comment]
I've noticed this change today but not sure if something is maybe going wrong on the Mastodon side. This post (https://darkfriend.social/@mmasnick.bsky.social@bsky.brid.gy/112638756668109759) includes a link (https://bsky.brid.gy/convert/ap/at://did:plc:dhkmbru2x57gpmzwufafq6ef/app.bsky.feed.post/3kv7nbofcnp2x) that seems like it should be to the quoted post (https://bsky.app/profile/dannagal.bsky.social/post/3kv7nbofcnp2x) but if it's meant to do a redirect that's just not firing, and instead I'm left looking at some JSON.

**Comment by snarfed** [comment]
Thanks for the nudge! You're right, that's not great, and not Mastodon's fault. I'll work on that.

**Comment by trwnh** [comment]
> It seems like current inbox delivery audience targeting could easily be
extended to send post Updates to all known quote post authors, which could
help with this.

This isn't really something that works unless you have an explicit idea of
a "quote post". Right now (and under e232), it's just a normal link like
any other. Having the linked resource change is not special, and adding a
notification mechanism for when any linked resource changes is pretty
firmly out-of-scope. This is the kind of thing that would be more likely to
be possible if there was an actual special collection of "quote posts",
which there currently is not (and I'm not sure there should be one, either
-- merely linking to a resource is not something that should be special
imo).

**Comment by snarfed** [comment]
Hi @trwnh! Sure, of course. We're squarely in e232-land here, and even then, you're right that there isn't necessarily a definition of a "quote post," even as a subset of linked resources.

I'm not seriously proposing or working on anything like update notifications for quote posts, it was just a casual hypothetical. _If I was,_ a first pass could be:
* A "quote post" is any object with an FEP-e232 attachment, notably one with `mediaType=application/ld+json ...` or `application/activity+json`.
* When a quote post is created, it should be delivered to the quoted object's owner's (`attributedTo` or `actor`) inbox, if available, along with other appropriate inboxes.
* AP servers track the objects that have quoted each object they own, similar to the existing `replies`, `shares`, etc collections. This could be a new externally-visible collection per object, or not.
* When an object is updated and its server sends an `Update` activity, it sends that activity to each quote post owner's inbox too.

I'm sure this has problems, but it's at least a straw man. I don't plan to take it farther than this though. 😁

**Comment by AtiusAmy** [comment]
> I'm still looking for a Misskey instance I can create an account on to test this. I've looked at https://misskey.io/ and https://quietplace.xyz/ , and neither of them can see @bsky.brid.gy@bsky.brid.gy, so I can't enable the bridge on them. makai.chaotic.ninja, which @temtemy mentioned above, doesn't have open signups. 
> 
> @AtiusAmy, any thoughts on where I can test this?

Misskey.art might work


https://misskey.art/@bsky.brid.gy@bsky.brid.gy

**Comment by AtiusAmy** [comment]
![Screenshot_20240619-082505_Firefox.png](https://github.com/snarfed/bridgy-fed/assets/136548767/1af56198-c619-4c95-8b2a-ac3dd5ad5f8b)

🎊🎉

**Comment by snarfed** [comment]
Woo, awesome! Thank you for testing!

**Comment by AtiusAmy** [comment]
> Woo, awesome! Thank you for testing!

I'm trying to see what a bsky quote would look on Sharkey, this post is not bridging tho...

https://bsky.app/profile/amyiscoolz.bsky.social/post/3kvakoqn3tc2m

**Comment by AtiusAmy** [comment]
![Screenshot_20240619-083416_Aria.png](https://github.com/snarfed/bridgy-fed/assets/136548767/1770c180-5406-4f43-b530-edd00271d3b2)

Viewed on Sharkey/Misskey :D

**Comment by snarfed** [comment]
Yes! I confirmed on Misskey just now too, https://quietplace.xyz/notes/9unjb8tho85n1rid

<img width="638" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/a719900a-3d2a-444d-9950-a8279b35cf04">

I expected Misskey to remove the RE: [URL] suffix, but oh well.

**Comment by AtiusAmy** [comment]
> Yes! I confirmed on Misskey just now too, https://quietplace.xyz/notes/9unjb8tho85n1rid
> 
> <img width="638" alt="image" src="https://github.com/snarfed/bridgy-fed/assets/778068/a719900a-3d2a-444d-9950-a8279b35cf04">
> 
> I expected Misskey to remove the RE: [URL] suffix, but oh well.

Akkoma and Threads look just like this on Misskey as well, so it should be fine

**Comment by snarfed** [comment]
OK, @hybridhavoc https://github.com/snarfed/bridgy-fed/issues/461#issuecomment-2176620836 should be fixed by ^ snarfed/granary@992e5b5, the RE: ... links should now work right in browsers.

Aaand I think that's it! Quote posts should now translate natively, both directions, if the quoted post's author has enabled the bridge. Otherwise, the post text will end with RE: ... with a link to the quoted post.

Phew, that took some work. Tentatively closing!

**Comment by dploeger** [comment]
Hmmm. This looks kinda weird now:

https://fed.brid.gy/r/https://bsky.app/profile/did:plc:vzeclfzjqfezzecw543cafaq/post/3kvc5vutiol2v

That link in there is some tech data about the original post I think.

**Comment by dploeger** [comment]
I'm seeing it on IceCubes and that renders links to other Mastodon posts inline for quotes. 

However, if they don't I would've expected the link after Re: to send me to the original Bluesky post, but that link only leads to some sort of JSON representation.

**Comment by snarfed** [comment]
Interesting! I expect IceCubes is using the Mastodon API to do that, which won't work with Bridgy Fed since it isn't Mastodon.

You're absolutely right that the link should work though! I thought I fixed that yesterday in https://github.com/snarfed/bridgy-fed/issues/461#issuecomment-2177434432 , but evidently that fix isn't always working. I'll look.

**Comment by dploeger** [comment]
Thanks a lot for your work!

**Comment by hybridhavoc** [comment]
That's interesting because the link posted above (https://fed.brid.gy/r/https://bsky.app/profile/did:plc:vzeclfzjqfezzecw543cafaq/post/3kvc5vutiol2v) works for me, sends me to the Bluesky post. But I've seen some other bridged Bluesky posts today (like this one https://darkfriend.social/@mmasnick.bsky.social@bsky.brid.gy/112644015357113938) that still just have the link to the JSON. 

![image](https://github.com/snarfed/bridgy-fed/assets/53440374/785cb1d3-cc76-455a-bb6d-146fc4bb128d)

My instance is Mastodon version 4.2.9. The quoted post is probably not itself bridged, so I don't know if that's actually the problem - that it's meant to display differently if it is bridged. If there is intended to be an embed, so far as I can tell that embed never appears for me.

**Comment by dploeger** [comment]
@hybridhavoc  yes it does. I'm sorry. I thought it would lead to a Mastodon website for it, but bridgy (intelligently) directly forwards to BlueSky. What I originally meant was the link INSIDE that post, after the RE (see the screenshot above)

**Comment by snarfed** [comment]
@hybridhavoc Mastodon doesn't support quote posts yet, so they won't show embeds at all. But you're also right that quoted posts that aren't bridged won't show as embeds even on platforms that do support them.

I think I've fixed the RE: links now. Example: https://bsky.app/profile/snarfed.bsky.social/post/3kvcdqj4vik2f => https://indieweb.social/@snarfed.bsky.social@bsky.brid.gy/112644716938090636 , screenshot below. Please post here again if you see any new bridged quote posts created after ~11:30a PT today that still have the old bad links!

![image](https://github.com/snarfed/bridgy-fed/assets/778068/c3a12ba9-7457-4813-a9e9-8c5c3380b0fa)

**Comment by mackuba** [comment]
Yup, looking good, thanks @snarfed! 👏

<img width="601" alt="Screen Shot 2024-06-19 at 20 51 04" src="https://github.com/snarfed/bridgy-fed/assets/28465/59f132a4-e089-4706-bd6b-e7ee53661d4e">

**Comment by snarfed** [comment]
Oh wow, nice, the link preview turns it into a hacky approximation of a quote post anyway, even on Mastodon. Love it.

**Comment by AtiusAmy** [comment]
https://bsky.app/profile/AmyIsCoolz.social.atiusamy.com.ap.brid.gy/post/3kvcmqcpywtt2

Does quote post not work with images on bsky or it can't fetch that specific post of mine (which was bridged afaik)

**Comment by AtiusAmy** [comment]
huh, when I quote post myself (without images) it does not seem to be bridged at all, I'll try again in a sec

**Comment by AtiusAmy** [comment]
*Phew* Not sure if I should create a new issue, but I narrowed down the problem!

https://social.atiusamy.com/@atiusamy

https://bsky.app/profile/atiusamy.social.atiusamy.com.ap.brid.gy

1. If I quote post a Bluesky user, it is bridged
2. If I quote post a Fediverse user (that is bridged) it will not be be bridged, however if I send an image in the quote post, it will be bridged just not the quote post.

**Comment by snarfed** [comment]
Thanks for the investigation @AtiusAmy! Will look.

**Comment by dploeger** [comment]
I can confirm that it works perfectly now. Thanks a lot for all your work, people! ♥️

**Comment by AtiusAmy** [comment]
> Thanks for the investigation @AtiusAmy! Will look.

Ok I found out something
![image](https://github.com/snarfed/bridgy-fed/assets/136548767/5a2dc765-4947-437b-bfb0-fc74eaa915a6)
![image](https://github.com/snarfed/bridgy-fed/assets/136548767/12c98970-b13c-415d-b656-b1b8c9dcd0be)

The amount of post is the same on bsky and AP, despite having less post on the BSKY side (You can count). Not sure what's going on, maybe it is bridged but the appview won't render?

**Comment by jamietanna** [comment]
Noticed that my first quote hasn't rendered properly - maybe my MF2 markup? 

Post: https://fed.brid.gy/r/https://www.jvt.me/mf2/2024/06/o0vwu/

![Screenshot_20240622-225708](https://github.com/snarfed/bridgy-fed/assets/3315059/6ae16c88-b4ff-4d5c-9ad2-f4fbae23f21f)

**Comment by AtiusAmy** [comment]
![image](https://github.com/snarfed/bridgy-fed/assets/136548767/b1c4bf1c-1e85-4ea8-91f0-f8153f03c18f)
Probably expected? Misskey supports native quote posting and they doesn't seem to show anything (aside from the weird unknown thing).

**Comment by snarfed** [comment]
Interesting! We haven't really tested web => AP quote posting yet; we clearly need to. @jamietanna I think your mf2 here is probably fine, the (initial) problem here is that BF interpreted it as an `article` quote, which doesn't trigger the rest of its quote post pipeline. Clearly wrong, the quoted URL is an AP `note`, so we can do better.

This issue is pretty long and hairy, I'll create a new one.

**Comment by snarfed** [comment]
(Oh and yeah @AtiusAmy is right, Mastodon itself doesn't support quote posts so you won't see it properly there anyway, but BF isn't quite working here regardless.)

**Comment by AtiusAmy** [comment]
> > Thanks for the investigation @AtiusAmy! Will look.
> 
> Ok I found out something
> ![image](https://github.com/snarfed/bridgy-fed/assets/136548767/5a2dc765-4947-437b-bfb0-fc74eaa915a6)
> ![image](https://github.com/snarfed/bridgy-fed/assets/136548767/12c98970-b13c-415d-b656-b1b8c9dcd0be)
> 
> The amount of post is the same on bsky and AP, despite having less post on the BSKY side (You can count). Not sure what's going on, maybe it is bridged but the appview won't render?

@snarfed Do you might know the reason of that (the post that aren't bridged are the quote post of AP user)

**Comment by snarfed** [comment]
@AtiusAmy I don't know for sure, but that's a good guess!

**Comment by snarfed** [comment]
OK @AtiusAmy I think I've fixed this issue, bridging fediverse quote posts of bridged native fediverse posts to Bluesky. (Phew, that's a mouthful.) Feel free to try again!

**Comment by AtiusAmy** [comment]
> OK @AtiusAmy I think I've fixed this issue, bridging fediverse quote posts of bridged native fediverse posts to Bluesky. (Phew, that's a mouthful.) Feel free to try again!

https://bsky.app/profile/atiusamy.social.atiusamy.com.ap.brid.gy/post/3kvrsg4xahiw2

Took a while to bridge but it works!

**Comment by snarfed** [comment]
Great! Yeah we were backed up for a bit, but we're through the backlog now.

**Comment by dploeger** [comment]
@snarfed  the problem with the link to the JSON representation instead of the quoted post somehow is back.

**Comment by snarfed** [comment]
@dploeger yes! sorry about that, tracked in #1154. I fixed it last night but had to roll back the fix. Attempt #2 is deploying now.

**Comment by silverpill** [comment]
Hello @snarfed ! I can confirm that your FEP-e232 implementation is correct.

You may also add `rel` property with value `https://misskey-hub.net/ns#_misskey_quote`. Some implementations do that to indicate that object link is a quote, though this is not required.
